src/elbo.jl

```diff
@@ -0,0 +1,209 @@
+exact solution. If there is no such solution, `Default()` either uses
+`Quadrature()` or `MonteCarlo()`, depending on the likelihood.
+
+N.B. the observation noise `fx.Σy` is assumed to be homoscedastic and
+uncorrelated - i.e. only `fx.Σy[1]` is used.
+
+[1] - Hensman, James, Alexander Matthews, and Zoubin Ghahramani. "Scalable
+variational Gaussian process classification." Artificial Intelligence and
+Statistics. PMLR, 2015.
+"""
+function elbo(
+    fx::FiniteGP,
+    y::AbstractVector{<:Real},
+    fz::FiniteGP,
+    q::AbstractMvNormal;
+    n_data=length(y),
+    method=Default()
+)
+    return _elbo(method, fx, y, fz, q, GaussianLikelihood(fx.Σy[1]), n_data)
+end
+
+
+"""
+    elbo(lfx::LatentFiniteGP, y::AbstractVector, fz::FiniteGP, q::AbstractMvNormal; n_data=length(y), method=Default())
+
+Compute the ELBO for a LatentGP with a possibly non-conjugate likelihood.
+"""
+function elbo(
+    lfx::LatentFiniteGP,
+    y::AbstractVector,
+    fz::FiniteGP,
+    q::AbstractMvNormal;
+    n_data=length(y),
+    method=Default()
+)
+    return _elbo(method, lfx.fx, y, fz, q, lfx.lik, n_data)
+end
+
+# Compute the common elements of the ELBO
+function _elbo(
+    method::ExpectationMethod,
+    fx::FiniteGP,
+    y::AbstractVector,
+    fz::FiniteGP,
+    q::AbstractMvNormal,
+    lik::ScalarLikelihood,
+    n_data::Integer
+)
+    post = approx_posterior(SVGP(), fz, q)
+    f_mean, f_var = mean_and_var(post, fx.x)
+    variational_exp = expected_loglik(method, y, f_mean, f_var, lik)
+
+    kl_term = StatsBase.kldivergence(q, fz)
+
+    n_batch = length(y)
+    scale = n_data / n_batch
+    return sum(variational_exp) * scale - kl_term
+end
+
+"""
+    expected_loglik(method, y, f_mean, f_var, lik)
+
+This function computes the expected log likelihood:
+
+```math
+    ∫ q(f) log p(y | f) df
+```
+where `p(y | f)` is the process likelihood.
+
+`q(f)` is an approximation to the latent function values `f` given by:
+```math
+    q(f) = ∫ p(f | u) q(u) du
+```
+where `q(u)` is the variational distribution over inducing points (see
+[`elbo`](@ref)). The marginal means and variances of `q(f)` are given by
+`f_mean` and `f_var` respectively.
+
+Where possible, this expectation is calculated in closed form. Otherwise, it is
+approximated using either Gauss-Hermite quadrature or Monte Carlo.
+
+# Extended help
+
+`q(f)` is assumed to be an `MvNormal` distribution and `p(y | f)` is assumed to
+have independent marginals such that only the marginals of `q(f)` are required.
+"""
+function expected_loglik end
+
+"""
+    expected_loglik(method::ExpectationMethod, y::AbstractVector, f_mean::AbstractVector, f_var::AbstractVector, lik::ScalarLikelihood)
+
+The expected log likelihood for a `ScalarLikelihood`, computed via `method`.
+Defaults to a closed form solution if it exists, otherwise defaults to
+Gauss-Hermite quadrature.
+"""
+function expected_loglik(
+    ::Default,
+    y::AbstractVector,
+    f_mean::AbstractVector,
+    f_var::AbstractVector,
```

Comment:
Just wondering aloud: would there be any (significant) runtime cost (overhead) in representing this as an `AbstractVector{<:Normal}` instead?

Reply:
Changed this to use `AbstractVector{<:Normal}` instead